Extract initial connection auth dance from the epics into module

jimmcgaw opened the pull request:
There are benefits to observables, arguably. I don't see any benefit to having this initial auth handshake with the server being contained in the epics and being done via observables.

jimmcgaw committed:
Extract initial connection auth dance from the epics into module

jimmcgaw commented on src/components/App.js:
```diff
@@ -67,10 +69,34 @@ class App extends Component {
+
+
+  onInitConnection(pincode='') {
+    const urlHash = document.location.hash + pincode;
```
Step 1. The first issue was that the very start of the observable auth flow had a dependency on the browser. This just injects the browser value into flow. When it's not the browser, ie a mobile device, it will come from a navigation screen parameter or something.

jimmcgaw commented on src/utils/sessions.js:
```diff
@@ -0,0 +1,60 @@
+  miniLock.crypto.getKeyPair(passphrase, email, (keyPair) => {
+    miniLock.session.keys = keyPair;
+    miniLock.session.keyPairReady = true;
+    let mID = miniLock.crypto.getMiniLockID(keyPair.publicKey);
```
Step 2: initial entry point, do the minilock stuff.

jimmcgaw commented on src/utils/sessions.js:
```diff
@@ -0,0 +1,60 @@
+    let mID = miniLock.crypto.getMiniLockID(keyPair.publicKey);
+    // 4. When we have keypair, create session on device
+    if (isNewPassphrase) {
+      createWebSession(passphrase);
```
Step 3: callback hook to do something on the device. In this case, just a browser, but might be something else on mobile, etc.

elimisteve replied:
@jimmcgaw In case more things are added to the body of `createWebSession()`, should we perhaps add `await` here (and make this callback we're creating here and passing to `miniLock.crypto.getKeyPair(...)` `async`)?

jimmcgaw replied:
I can add that.

elimisteve replied:
Gracias :grin: :man_dancing:

jimmcgaw commented on src/utils/sessions.js:
```diff
@@ -0,0 +1,60 @@
+          authToken,
+          secretKey,
+          mID,
+          isNewRoom: isNewPassphrase,
```
Step 4. Do the auth dance, get the token, and send that to the init connection action in redux.

jimmcgaw commented on src/actions/chatActions.js:
```diff
@@ -13,8 +13,13 @@ export const CHAT_USERNAME_SET = 'CHAT_USERNAME_SET';
+  ({
+    type: CHAT_INIT_CONNECTION,
+    authToken, secretKey, mID, isNewRoom
+  });
```
Step 5. This is the action dispatched by Step 4, when doing the auth. Now we enter the observable flow, when it receives this action type.

jimmcgaw commented on src/epics/chatEpics.js:
```diff
@@ -73,98 +67,34 @@ const suggestionEpic = (action$) =>
+};
 
 const initConnectionEpic = (action$) =>
   action$.ofType(CHAT_INIT_CONNECTION)
```
Step 6: This listens for this action to be triggered. We enter this flow with the action parameters. We pass those, first thing, to the initChatHandlerConnection method in here.

jimmcgaw commented on src/epics/chatEpics.js:
```diff
@@ -73,98 +67,34 @@ const suggestionEpic = (action$) =>
+    authToken,
+    secretKey: secretKey,
+    mID
   });
```
Step 7. Initialize the chat handler's connection. Since `chatHandler` is a shared global variable in this module, this is as far as we can go, without ripping all of the other observable stuff out of here too. This returns an observable.

elimisteve replied:
So much more readable! Bravo 👏

jimmcgaw commented on src/epics/chatEpics.js:
```diff
@@ -73,98 +67,34 @@ const suggestionEpic = (action$) =>
-    )
+    .mergeMap(initChatHandlerConnection)
+    .mergeMap((isNewRoom) =>
+      Observable.merge(
```
Step 8: with `chatHandler` initialized and connected, do all the same stuff we did before.

jimmcgaw commented:
@elimisteve There will be a conflict between this and other open PR that reworks major chunks of the frontend. In terms of ease of resolving conflicts, I'd rather merge the other one first.

elimisteve commented on src/components/App.js:
```diff
@@ -67,10 +69,34 @@ class App extends Component {
   connectIfNeeded() {
     if (!this.props.pincodeRequired && this.props.shouldConnect) {
-      this.props.initConnection();
+      // this.props.initConnection();
```
Can this line be removed?

jimmcgaw replied:
Ah, yep. Last push was some minor cleanup of three lines, including removing this.

elimisteve commented on src/utils/sessions.js:
```diff
@@ -0,0 +1,60 @@
+    headers: {
+      'X-Minilock-Id': mID
+    }
+  });
```
`fetch` ftw 👌

jimmcgaw commented on src/utils/sessions.js:
```diff
@@ -0,0 +1,60 @@
+    let mID = miniLock.crypto.getMiniLockID(keyPair.publicKey);
+    // 4. When we have keypair, create session on device
+    if (isNewPassphrase) {
+      await createWebSession(passphrase);
```
Step 3. Etc

jimmcgaw commented on src/components/chat/ChatContainer.js:
```diff
@@ -37,8 +37,6 @@ ChatContainer.propTypes = {
 const mapStateToProps = (reduxState) => {
   return {
-    messages: reduxState.chat.messages,
-    username: reduxState.chat.username,
```
Small clean up from the state PR. Piggybacking onto this.

jimmcgaw commented on src/components/App.js:
```diff
@@ -59,12 +83,16 @@ class App extends Component {
+
+  onSetUsername = (username) => {
+    this.props.setUsername(username);
+    this.onToggleModalVisibility('username', false);
```
Snuck in on bad merge with develop. Will remove.

elimisteve replied:
Feel free to merge this in once you do! Lookin' good :sunglasses:

jimmcgaw committed:
Remove cruft from develop merge

jimmcgaw committed:
Fix subtle bug

jimmcgaw commented:
@elimisteve I just added a commit to this, and you'll want to take a look. I was seeing a subtle bug. Sometimes, when I connected, my messages were double posting into the message box. On page reload, the duplicate messages disappeared.

Why? The app, via redux state, is one big finite state machine. Dispatching actions handles the state transitions. (You're probably aware of this. Giving context.)

After initializing the chat, the app goes through these state transitions:

`(shouldConnect) == initConnection ==> (connecting) == connectionInitiated ==> (connected)`

The app continuously checks its own state (at least, on every re-render) to see if it's in shouldConnect state. If so, it initiates an initConnection transition. Previously, the init connection updated the state to "connecting" immediately, so any re-renders wouldn't trigger a duplicate connection. But, now that we moved the initConnection out of the observable flow and into its own module, there was a significant delay between init'ing the connection process and the state being updated. The state now gets updated at the very END of the flow, once the auth handshake has completed.

This caused a subtle race condition. If the user updated the state of the app during the auth handshake (say, by setting their username) it would look and find the app in the `shouldConnet` state, even though the connection had been initiated. So another auth / connection was initiated. The result? You'd end up with double connections handling requests on the `chatHandler`. This would result in double messages being appended to the local `messages` list, but not double requests to the server. (So a page reload would remove the duplicate messages.)

I updated the state machine and added an additional state, and transition. It now looks like this:

`(shouldConnect) == initAuth ==> (authenticating) == initConnection ==> (connecting) == connectionInitiated ==> (connected)`

`initAuth` action dispatch immediately transitions the app out of `shouldConnect` state, so even if there are re-renders, it won't initiate a second connection.

jimmcgaw commented on src/components/App.js:
```diff
@@ -123,8 +142,18 @@ const mapStateToProps = (reduxState) => {
   return {
     initChat: () => dispatch(initChat()),
-    initConnection: (pincode) => dispatch(initConnection(pincode)),
-    initApplication: () => dispatch(initApplication()),
```
`initApplication` looks like it's no longer used, or even defined, anywhere.

elimisteve commented:
Excellent; thanks for adding the `connecting` state

elimisteve commented:
Or, rather, the `CHAT_INIT_AUTH` state.
